src/v/cluster/members_backend.cc

```diff
@@ -0,0 +1,310 @@
+              case members_manager::node_update_type::decommissioned:
+                  handle_decommissioned(update);
+                  break;
+              case members_manager::node_update_type::recommissioned:
+                  handle_recommissioned(update);
+                  break;
+              default:
+                  return;
+              }
+              _new_updates.signal();
+          });
+      })
+      .handle_exception([](const std::exception_ptr& e) {
+          vlog(clusterlog.trace, "error waiting for members updates - {}", e);
+      });
+}
+void members_backend::start_reconciliation_loop() {
+    (void)ss::with_gate(_bg, [this] {
+        return reconcile().finally([this] {
+            if (!_as.local().abort_requested()) {
+                _retry_timer.arm(_retry_timeout);
+            }
+        });
+    }).handle_exception([](const std::exception_ptr& e) {
+        // just log an exception, will retry
+        vlog(
+          clusterlog.trace,
+          "error encountered while handling cluster state reconciliation - {}",
+          e);
+    });
+}
+
+ss::future<> members_backend::reconcile() {
+    // if nothing to do, wait
+    co_await _new_updates.wait([this] {
+        return !_reallocations.empty()
+               || !_members.local().get_decommissioned().empty();
+    });
+
+    auto u = co_await _lock.get_units();
+
+    for (auto& meta : _reallocations) {
+        mark_done_as_finished(meta);
+    }
+
+    if (!_raft0->is_leader()) {
+        co_return;
+    }
+
+    // execute reallocations
+    co_await ss::parallel_for_each(
+      _reallocations,
+      [this](reallocation_meta& meta) { return reallocate_replica_set(meta); });
+
+    // remove finished meta
+    std::erase_if(_reallocations, [](const reallocation_meta& meta) {
+        return meta.state == reallocation_state::finished;
+    });
+
+    // remove those decommissioned nodes which doesn't have any pending
+    // reallocations
+    std::vector<model::node_id> to_remove;
+    for (auto& id : _members.local().get_decommissioned()) {
+        auto has_active_reallocations = std::any_of(
+          _reallocations.cbegin(),
+          _reallocations.cend(),
+          [id](const reallocation_meta& meta) { return id == meta.update.id; });
+
+        if (!has_active_reallocations && _allocator.local().is_empty(id)) {
+            to_remove.push_back(id);
+        }
+    }
+
+    // no nodes to remove from cluster
+    if (to_remove.empty()) {
+        co_return;
+    }
+    vlog(clusterlog.info, "removing decommissioned nodes: {}", to_remove);
+    // we are not interested in the result as we check
+    // configuration in next step
+    co_await _raft0->remove_members(to_remove, model::revision_id{0})
+      .discard_result();
+
+    auto cfg = _raft0->config();
+    // we have to wait until raft0 configuration will be fully up to date.
+    if (cfg.type() != raft::configuration_type::simple) {
+        co_return;
+    }
```

Comment:
exactly, it does, after configuration update `consensus` enters joint consensus state. There is no point in checking configuration during that time since this state is transient. After update will successfully finish (in next retry) code checks if the node was actually removed from `memebrs_table`